roomserver/internal/perform/perform_upgrade.go

```diff
@@ -0,0 +1,691 @@
+	newCreateContent := map[string]interface{}{}
+	if err := json.Unmarshal(oldCreateEvent.Content(), &newCreateContent); err != nil {
+		util.GetLogger(ctx).WithError(err).Error()
+		return nil, &api.PerformError{
+			Msg: "Create event content was invalid",
+		}
+	}
+	newCreateContent["creator"] = userID
+	newCreateContent["room_version"] = newVersion
+	newCreateContent["predecessor"] = gomatrixserverlib.PreviousRoom{
+		EventID: tombstoneEvent.EventID(),
+		RoomID:  roomID,
+	}
+	newCreateEvent := fledglingEvent{
+		Type:    gomatrixserverlib.MRoomCreate,
+		Content: newCreateContent,
+	}
+
+	membershipContent := gomatrixserverlib.MemberContent{}
+	if err := json.Unmarshal(oldMembershipEvent.Content(), &membershipContent); err != nil {
+		util.GetLogger(ctx).WithError(err).Error()
+		return nil, &api.PerformError{
+			Msg: "Membership event content was invalid",
+		}
+	}
+	membershipContent.Membership = gomatrixserverlib.Join
+	membershipEvent := fledglingEvent{
+		Type:     gomatrixserverlib.MRoomMember,
+		StateKey: userID,
+		Content:  membershipContent,
+	}
+
+	powerLevelContent, err := oldPowerLevelsEvent.PowerLevels()
+	if err != nil {
+		util.GetLogger(ctx).WithError(err).Error()
+		return nil, &api.PerformError{
+			Msg: "Power level event content was invalid",
+		}
+	}
+
+	// We might need to temporarily give ourselves a higher power level
+	// than we had in the old room in order to be able to send all of
+	// the relevant state events. This function will return whether we
+	// had to override the power level events or not — if we did, we
+	// need to send the original power levels again later on.
+	tempPowerLevelsEvent, powerLevelsOverridden := createTemporaryPowerLevels(powerLevelContent, userID)
+
+	joinRulesContent, err := oldJoinRulesEvent.JoinRule()
+	if err != nil {
+		return nil, &api.PerformError{
+			Msg: "Join rules event content was invalid",
+		}
+	}
+	newJoinRulesEvent := fledglingEvent{
+		Type: gomatrixserverlib.MRoomJoinRules,
+		Content: map[string]interface{}{
+			"join_rule": joinRulesContent,
+		},
+	}
+
+	eventsToMake := make([]fledglingEvent, 0, len(state))
+	eventsToMake = append(eventsToMake, newCreateEvent, membershipEvent, tempPowerLevelsEvent, newJoinRulesEvent)
+
+	if _, ok := state[gomatrixserverlib.StateKeyTuple{EventType: gomatrixserverlib.MRoomGuestAccess, StateKey: ""}]; !ok {
+		// Appease sytest, as it expects a guest access event for some reason.
+		eventsToMake = append(eventsToMake, fledglingEvent{
+			Type: gomatrixserverlib.MRoomGuestAccess,
+			Content: map[string]string{
+				"guest_access": "forbidden",
+			},
+		})
+	}
+
+	// Duplicate all of the old state events into the new room.
+	for tuple, event := range state {
+		if _, ok := override[tuple]; ok {
+			// Don't duplicate events we have overridden already. They
+			// are already in `eventsToMake`.
+			continue
+		}
+		if membership, merr := event.Membership(); merr == nil && membership != gomatrixserverlib.Ban {
```

Comment:
It's icky that we're doing exactly the same check in two places. It would be better if we did a single pass over the old state and bucket them into "stuff to send" and "stuff to ignore".

Reply:
I think the second check is actually not needed, so I'll remove it.